man/shutdown.8
.TH ninit-shutdown 8 "Dec 28, 2009"
.SH NAME
ninit-shutdown \- shutdown the ninit init system.
.SH SYNOPSIS
.B ninit-shutdown 
\-[\fIrhoqvstmST\fR] [\-E /path/to/program [arg1 arg2 ...]]

.SH DESCRIPTION
.B ninit-shutdown
tries to properly shutdown your system with ninit.

.SH USAGE
.TP 9
\-r
reboot after shutdown
.TP
\-h
halt after shutdown
.TP
\-o
power-off after shutdown (default)
.TP
\-q
quiet mode;  ignores SIGINT signal;  Ctrl-C does't work!
.TP
\-v
be verbose
.TP
\-m
only shutdown the ninit-part and exit
.TP
\-s \fIsecs\fR 
starting delay
.TP
\-t \fIsecs\fR
delay between SIGTERM and SIGKILL
.TP
\-E \fIprog\fR
execute \fIprog\fR after KILLALL;  must be the last option!
.TP
\-T \fIsecs\fR
if \fIsecs\fR is nonzero shutdown twice ninit part
.TP
\-S \fIabcd\fR
skip to shutdown the service \fIabcd\fR

.SH NOTES
If the flags -m and -E are off then the program starts the service
.I halt 
or 
.I reboot 
depending of the flags -h -o  and -r respectively.

If the flags -E and -m are on then 
.B ninit-shutdown 
execve(2)s the
\fIprog\fR after shutting down the ninit-part.

If the flag -E is on and the flag -m is off then 
.B ninit-shutdown
execve(2)s
the \fIprog\fR after sending SIGTERM and SIGKILL to all processes.

.B ninit-shutdown
clears respawn/cron flags on all services
including those skipped with -S flag.

.
.SH "ENVIRON"
.B ninit-shutdown
uses the variables 
.B NINIT_MEMORY 
and 
.B NINIT_HOME. 

.SH AUTHOR
.B ninit 
was written by Nikola Vladov and can be downloaded from
.br
.I http://riemann.frmi.uni-sofia.bg/ninit/

Parts of
.B ninit-shutdown
were contributed by Bernd Wachter.
.br
Parts of this manpage was written by Erich Schubert.

.SH "SEE ALSO"
ninit(8), nsvc(8), execve(2), nkillall(8)


CHANGES
0.14:   2010-01-19 00:27:01
  added nkillall.8  
  gid is not default anymore.  see ninitfeatures.h
  uid supports 22:33:100:200:300
  nkillall: \e escape
  shutdown: set cron off
  nkillall: signals with letters like nsvc
  shutdown and nkillall: makes more sync(2), flag -q
  added THANKS ;-)
  added helpers: remove procfs;  thanks to them snvc is smaller
  nsvc options -onumber -unumber; option -E
  Cron has option -C
  /etc/ninit/.nsvc_help
  nsvc, shutdown, reload uses variables NINIT_MEMORY and NINIT_HOME
  nsvc: option -S
  shutdown: option -s
  manpages for all programs in /sbin ;-)
  update manpages (Boris Grozev)
  Makefile: man_txt
  serdo.c: read only one file again
  better sleep in open_inout.h
  serdo.c improvments; buildin: killall5, simple echo
  nsvc.c: -C+number
  scripts/conf
  ninit.8: NONROOT USAGE, using /sbin/init with ninit
  ninit-scan: (Wayne Marshall)

0.13.7:   2009-12-16 10:47:50
  bootlog.c (flush_root)
  0.13.1.4 renamed to 0.13.2
  service start 'end' (if it exixts; X_OK) after finishing
  fixed a bug in shutdown.c (thx to Jan)
  With flag -s ninit-shutdown don't start halt and reboot services
  shutdown.c:  improvements; flags -S -q
  shutdown.c:  removed flags -s -q; new flags -E -v -T
  added nkillall
  shutdown.c:  options -m and -E work together
  nkillall print escape like echo

0.13.1:   2009-06-27 12:59:38
  updated bootlog
  Makefile logs 'make tests' in tests_log using boolog
  fixing quotes in serdo

0.13:   2009-01-10 18:11:42
  it's possible to set parameters in wait.  example: some_service:180:3
  install-bin print help and string errors
  better sync mode.  maxsync is removed.  echo 200 > sync
  updated scripts/ninit_test.sh
  comments (#) in depends, params, environ, wait, softlimit
  install-bin (verbose mode)
  nsvc -V
  errmsg_put.c; buffer_*.c
  setup, rsetup, sys-rsetup starts with args: $1=service $2=service_pid
  EXTRACT_* trick (automatic create header files)
  stuct process is 20 bytes on x86_64 (see struct_root.h)
  changed BIFFER_INIT macro
  utmp_io.c
  Makefile: install_other
  serdo is intslled only with: make install_other
  t_write.h uses PIPE_BUF (thx Laurent Bercot)
  nsvc.8 improvements
  ninit_test.sh creates services in ./etc/ninit/
  ninit-huge
  added printf.c and ninit.spec
  make FLAG_DEBUG=no
  removed flag X_OK in some syscalls access.
  better addprocess

0.12.1:   2008-01-03 17:15:08
  ninit-depends.  convert directory to file (depends.dir -> depends)
  removed unused stat.h in headers
  updated serdo (static)
  two environ vars: NINIT_HOME, INIT_HOME.  (Thanks to Stamatis Mitrofanis)
  ninit logo: /etc/ninit/.sync
  install-bin uses chown32 instead chown on some systems
  err.c and err_b.c uses the macros va_start and va_arg
  fu.c (u &= 0xffffffff)

0.12:
  almost all arrays are moved in stack.  small data and bss sections.
  assembler functions (i386) for some DJB functions.
  modifications in Makefile.
  100 Euro security guarantee.

0.11.2:
  ninit-reload accept agument -e (change environ) -E file_env
  program ~/sys/update
  service flags: pause, pause-wait
  service flag cron:  a:b[:c]
    can contain many lines.
  flag -K for nsvc.
  pidfilehack --> ninit-pidfile.
  make i386 -- build static daemons (don't use dietlibc)
  Makefile is smaller and simpler
  gcc-4.1.2 works now.
  install-bin
  stuct mem in initialized in the main program.

0.11.1:
  ninit-reload accepts args (-r3 -r5 -r33 ...)
  ninit-inittab sets the variables NINIT_RUNLEVEL, INIT_VERSION
  if /etc/minit/{in|out} FIFO's exist then make symbolic links.
  new run flags: sys-rsetup, pidfile
  if the service sysvinit exist, ninit create /dev/initctl 
  ninit-reload and nsvc checks for environ NINIT_HOME
  nsvc -[drR] ALL - change all services.
    Change service ALL with: nsvc -[drR] ops ALL
  new program: ninit-sysvinit.  configuration file: sysvinit-timeout
    fimeout[:fork-mode].  example:  echo 600:0 > sysvinit-timeout
  ninit catch signal SIGPWR
  new run flags: alarm, maxsync
  nsvc output is different
  shell script ninit-service
  new program run-wait

0.11:
  Sun Jun 17 22:08:22 EEST 2007
  I decided to rename all without changes.
  The name ninit must be spelled nano-init.

0.9.11:
  rewritten by Nikola Vladov

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  See minit CHANGES - http://www.fefe.de/minit


shutdown.c
/*
 * Notes:
 * - uncomment `#define USE_INIT' below if you want to use shutdown
 *   with ninit. If defined, shutdown will try to bring down the services 
 *   halt (for -h or -o) or reboot (-r) before killing all other processes.
 * - If you do not use ninit shutdown will bring your system down similar
 *   to SysV-Inits shutdown with -n
 */

#include <sys/types.h>
#include <unistd.h>
#include <sys/reboot.h>
#include <signal.h>
#include <fcntl.h>
#include <time.h>
#include <alloca.h>
#include <sys/wait.h>
#include <errno.h>
#include <stdarg.h>
#include "ninitfeatures.h"
#include "uid.h"
#include "check_opt.h"
#define Y(...) exec_cmd(__VA_ARGS__, 0)

#define USE_INIT

static char *me;
static int cfg_verbose;

void sighandler(int sig) { (void)sig; }

static void exec_argv(char **argv) {
  execve(argv[0], argv, environ);
  errmsg_iam(me);
  carp("execve ", argv[0]," failed"); 
}

static void exec_cmd(char *cmd, ...) {
  char *argv[25], **arg=argv; /* XXX can overflow */
  const char *s;
  va_list a;
  pid_t pid;
  int count_args = 0;
  
  *arg++ = cmd;
  count_args++;
  va_start(a,cmd);
  do {
    s=va_arg(a,const char*);
    *arg++ = (char *)s;
    count_args++;
  } while (s && count_args < 25);

  while ((pid = fork()) < 0) nano_sleep(0,500000000);
  if (pid == 0) {
    exec_argv(argv);
    _exit(0);
  } else
    while (pid != waitpid(-1,0,0)); /* also collect zombies */
}

static void deep_sleep(int t) {
  time_t deedline = time(0) + t;
  if (t<=0) return;
  if (cfg_verbose) msg("sleeping ", fu(t), "...");
  while (1) {
    nano_sleep(1,0);
    if (time(0) >= deedline) break;
  }
}

static void die_xx(char *s) { errmsg_iam(me); carp(s); _exit(1); }

#ifdef USE_INIT
#include "ninit.h"

static int infd, outfd;
static char *buf, do_update, *nsvc_other;
static unsigned int cfg_twice;
static pid_t my_pid;

#include "get_services.h"
#include "open_inout.h"
#undef CLEAN_SERVICE
#include "tryservice_nsvc.h"

static void try_srv(char *service, char *type) {
  if (tryservice(service, type[0]))
    carp("Could not ", type, " service: ", service);
}

void ninit_servicesDown(char **skip) {
  int idx;
  for (idx=0; idx<=maxprocess; idx++) {
    char *what = "not running";
    pid_t pid = root[idx].pid;
    char *name = root_name(idx);

    if (root[idx].pr_respawn) try_srv(name, "respawn off");
    if (root[idx].cron) {
      nsvc_other = "0";
      try_srv(name, "cron off");
      nsvc_other = 0;
    }
    if (pid > 1) {
      char **x;
      for (x=skip; *x; x++)
	if (!str_diff(*x, name)) { pid=my_pid; what="skipped"; break; }

      if (my_pid != pid) {
	what = "done";
	if (kill(pid, SIGTERM) == 0) kill(pid, SIGCONT);
	else what = "failed";
      }
    }
    if (cfg_verbose) msg("\t--> ", name, "\t\t", what);
  }
}

void ninit_shutdown(unsigned int twice, char **skip) {
  char *x;
  INIT_ENV_GET_HOME(x,"NINIT_HOME","INIT_HOME");
  if (x==0) x=INITROOT;
 again:
  open_inout(x);
  msg("Shutting down ninit services");  

  get_services();
  ninit_servicesDown(skip);

  if (twice) {
    close(infd);
    close(outfd);
    mem.l = 0;
    mem.r = mem.a;
    root = 0;
    maxprocess = -1;
    deep_sleep(twice);
    twice = 0;
    goto again;
  }
}
#endif

static void printUsage() {
  carp("usage: ",me," -[rhoqvt"
#ifdef USE_INIT
       "mST"
#endif
       "] [-E program [arg(s)]]\n"
       "\t-r:\treboot after shutdown\n"
       "\t-h:\thalt after shutdown\n"
       "\t-o:\tpower-off after shutdown\n"
       "\t-q:\tquiet mode; ignores SIGINT signal\n"
#ifdef USE_INIT
       "\t-m:\tonly shutdown the ninit part\n"
#endif
       "\t-v:\tbe verbose\n"
       "\t-E prog:\texecute prog after KILLALL;"
       "  must be last option!\n"
       "\t-s secs:\tstarting delay\n"
       "\t-t secs:\tdelay between SIGTERM and SIGKILL\n"
#ifdef USE_INIT
       "\t-T secs:\tif secs is nonzero shutdown twice ninit part\n"
       "\t-S abcd:\tskip to shutdown the service abcd\n"
#endif
       );
  _exit(1);
}

int main(int argc, char **argv) {
  char *x;
  unsigned int cfg_downlevel=2; /* 0:1:2 == reboot:halt:poweroff */
  unsigned int cfg_delay = 3;
  static unsigned int cfg_sleep;
  static int cfg_ninitonly, cfg_quiet;
  static char **prog;

#ifdef USE_INIT
  static char **skip, **skp;
  skip = alloca((argc+5) * sizeof(char *));
  skp = ++skip;

  mem.a = INIT_ALLOC_BUFFER;
  buf = alloca(BUFFER_TMP_LEN);
#endif

  if (getuid() != 0) die_xx("You are not root, go away!");
  me = argv[0];
  if (argc<2) printUsage();

  /* parse commandline options */
  argv++;
  while (argv[0] && argv[0][0]=='-') {
    char *y=argv[0] + 1;
    for (; *y; y++) {
      switch(*y) {
      case 'q': /* ignore SIGINT */	++cfg_quiet; break;
      case 'v': /* verbose mode */	++cfg_verbose; break;
      case 'r': /* reboot */	cfg_downlevel = 0; break;
      case 'h': /* halt */	cfg_downlevel = 1; break;
      case 'o': /* poweroff */	cfg_downlevel = 2; break;
      case 't': /* delay between SIGTERM and SIGKILL */
	chk_opt(argv,x); cfg_delay = x_atoi(x); goto again;
      case 's': /* sleep before shutdown services */
	chk_opt(argv,x); cfg_sleep = x_atoi(x); goto again;
#ifdef USE_INIT
      case 'm': /* exit after ninit down */
	cfg_ninitonly = 1; break;
      case 'T': /* delay between two services down */
	chk_opt(argv,x); cfg_twice = x_atoi(x); goto again;
      case 'S': /* skip services */
	chk_opt(argv,x); *skp++ = x; goto again;
      case 'E': /* execve prog arg(s) */
	chk_opt(argv,x); prog = argv; *prog = x; goto last_option;
#endif
      default:
	printUsage();
      }
    }
  again:
    argv++;
  }

#ifdef USE_INIT
 last_option:
  *skp = 0;
  *--skip = cfg_downlevel ? "halt" : "reboot";
#endif

  opendevconsole();

  if (cfg_ninitonly == 0) {
    switch (cfg_downlevel) {
    case 0: x="reboot"; break;
    case 1: x="halt"; break;
    case 2: x="power-off";
    }
    msg(me, "\007: System is going down for ",x);
  }
  sync();

  /* catch some signals; getting killed after killing the controlling terminal
   * wouldn't be such a great thing... 
   */
  set_sa(SIGQUIT); set_sa(SIGCHLD); set_sa(SIGHUP );
  set_sa(SIGTSTP); set_sa(SIGTTIN); set_sa(SIGTTOU);

  if (cfg_quiet) set_sa(SIGINT);
  deep_sleep(cfg_sleep);

#ifdef USE_INIT
  x = env_get("NINIT_MEMORY");
  if (x) mem.a = atoulong(x);
  mem.r=mem.a;
  if (mem.a==0 || mem.a > 64000 || (mem.x=alloca(mem.a))==0) _exit(1);
  my_pid = getpid();

  sync();

  ninit_shutdown(cfg_twice, skip);
  if (cfg_ninitonly == 0 && prog == 0) {
    msg("Starting service: ", *skip);
    try_srv(*skip, "start");
  }
  close(infd);
  close(outfd);
  sync();

  if (cfg_ninitonly) {
    if (prog) exec_argv(prog);
    _exit(0);
  }
  deep_sleep(cfg_delay);
#endif

  /* kill all processes still left */
  sync();
  msg("Sending all processes the ", "TERM signal...");
  kill(-1, SIGTERM);
  deep_sleep(cfg_delay);

  sync();
  msg("Sending all processes the ", "KILL signal...");
  kill(-1, SIGKILL);

  if (prog) { sync(); exec_argv(prog); }
  sync();

  /* maximal 20 arguments !!! */
  if (!access("/sbin/quotaoff", X_OK)) Y("/sbin/quotaoff", "-a");
  Y("/sbin/swapoff", "-a");	sync();
  Y("/bin/umount", "-a");
  Y("/bin/mount", "-o", "remount,ro", "/");
  sync();

  /* and finally reboot, halt or power-off the system */
  switch (cfg_downlevel) {
  case 0: cfg_downlevel = RB_AUTOBOOT; break;
  case 1: cfg_downlevel = RB_HALT_SYSTEM; break;
  case 2: cfg_downlevel = RB_POWER_OFF;
  }

  set_reboot(cfg_downlevel);
  return 0;
}
